base/iobuffer.jl

```diff
@@ -536,38 +739,85 @@ _unsafe_take!(io::IOBuffer) =
         io.size - io.offset)
 
 function write(to::IO, from::GenericIOBuffer)
-    written::Int = bytesavailable(from)
+    available = bytesavailable(from)
+    # If they're the same buffer, we need to special case it since the buffer
+    # is being mutated twice in the write call.
     if to === from
-        from.ptr = from.size + 1
+        if from.append
+            # If from.append, we copy data to from.size + 1.
+            # Note that this ensureroom might switch the data buffer, or update
+            # the fields like from.size
+            ensureroom(from, available)
+            data = from.data
+            size = from.size
+            existing_space = lastindex(data) - min(from.maxsize, size)
+            to_write = min(existing_space, available)
+            iszero(to_write) && return 0
+            GC.@preserve from unsafe_copyto!(data, size + 1, data, from.ptr, to_write)
+            from.size = size + to_write
+            from.ptr += to_write
+            return to_write
+        else
+            # Else, we copy data to itself, meaning we don't need to copy the data at all.
+            from.ptr = from.size + 1
+            return available
+        end
     else
-        written = GC.@preserve from unsafe_write(to, pointer(from.data, from.ptr), UInt(written))
+        written = GC.@preserve from unsafe_write(to, pointer(from.data, from.ptr), UInt(available))
         from.ptr += written
     end
-    return written
+    return available
 end
 
 function unsafe_write(to::GenericIOBuffer, p::Ptr{UInt8}, nb::UInt)
     ensureroom(to, nb)
-    ptr = (to.append ? to.size+1 : to.ptr)
-    written = Int(min(nb, Int(length(to.data))::Int - ptr + 1))
-    towrite = written
-    d = to.data
-    while towrite > 0
-        @inbounds d[ptr] = unsafe_load(p)
-        ptr += 1
+    size = to.size
+    append = to.append
+    ptr = append ? size+1 : to.ptr
+    data = to.data
+    to_write = min(nb % Int, min(Int(length(data))::Int, to.maxsize) - ptr + 1)
```

Comment:
Comparison should be done as unsigned to avoid large nb becoming a negative value, since it is the only number here that we don't know has a reasonable value
```suggestion
    to_write = min(nb, (min(Int(length(data))::Int, to.maxsize) - ptr + 1) % UInt) % Int
```